firestore/src/Connector.gs

```diff
@@ -0,0 +1,288 @@
+/*
+Copyright 2018 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/** 
+ * Contains core functionality for Data Studio connector. 
+ *
+ * @constructor
+ */
+function Connector(enableLogging) {
+
+  /** @const */
+  this.FIELD_TYPES = [
+    'STRING',
+    'NUMBER',
+    'BOOLEAN',
+    'TIMESTAMP'
+  ];
+
+  /** @const */
+  this.TIMESTAMP_SEMANTICS = {
+    conceptType: 'DIMENSION',
+    semanticGroup: 'DATE_AND_TIME',
+    semanticType: 'YEAR_MONTH_DAY_HOUR'
+  };
+  
+  /** @const */
+  this.cloud = new GoogleCloud();
+  
+  /** @const */
+  this.enableLogging = enableLogging;
+}
+
+
+/**
+ * Stringifies parameters and responses for a given function and logs them to
+ * Stackdriver.
+ *
+ * @param {string} functionName Function to be logged and executed.
+ * @param {Object} parameter Parameter for the `functionName` function.
+ * @returns {any} Returns the response of `functionName` function.
+ */
+Connector.prototype.logAndExecute = function(functionName, parameter) {
+  if (this.enableLogging) {
+    const paramString = JSON.stringify(parameter, null, 2);
+    console.log([functionName, 'request', paramString]);
+  }
+
+  const returnObject = this[functionName](parameter);
+
+  if (this.enableLogging) {
+    const returnString = JSON.stringify(returnObject, null, 2);
+    console.log([functionName, 'response', returnString]);
+  }
+
+  return returnObject;
+};
+
+
+/** 
+ * Predefined Data Studio function for specifying connector configuration.
+ *
+ * @param {Object} request Config request parameters.
+ * @return {Object} Connector configuration to be displayed to the user.
+ */
+Connector.prototype.getConfig = function(request) {
+  const projects = this.cloud.listCloudProjects();
+  const options = [];
+  projects.forEach(function(project) {
```

Comment:
This logic would be a bit simpler if you did  `const options = projects.map(...)`